pkg/elf/elf_test.go

```diff
@@ -0,0 +1,216 @@
+		elfChangeErr error
+	}
+	testOptions := []testOption{
+		{
+			description: "test direct copy",
+			elfValid:    validOptions,
+		},
+		{
+			description:  "test constant substitution 1",
+			key:          "FOO",
+			kind:         symbolUint32,
+			intValue:     42,
+			elfValid:     validOptions,
+			elfChangeErr: errDifferentFiles,
+		},
+		{
+			description:  "test constant substitution 2",
+			key:          "BAR",
+			kind:         symbolUint32,
+			intValue:     42,
+			elfValid:     validOptions,
+			elfChangeErr: errDifferentFiles,
+		},
+		{
+			description:  "test map name substitution name",
+			key:          "test_cilium_calls_4278124286",
+			kind:         symbolString,
+			strValue:     "test_cilium_calls_0000000042",
+			elfValid:     validOptions,
+			elfChangeErr: errDifferentFiles,
+		},
+		{
+			description: "test invalid map name substitution",
+			key:         "test_cilium_calls_4278124286",
+			kind:        symbolString,
+			strValue:    "test_cilium_calls_00",
+			elfValid:    notValidOptions,
+		},
+	}
+
+	for i := 1; i <= 2; i++ {
+		testOptions = append(testOptions, testOption{
+			description:  fmt.Sprintf("test mac substitution %d", i),
+			key:          fmt.Sprintf("LOCAL_MAC_%d", i),
+			kind:         symbolUint32,
+			intValue:     42,
+			elfValid:     validOptions,
+			elfChangeErr: errDifferentFiles,
+		})
+	}
+
+	for i, test := range testOptions {
+		c.Logf("%s", test.description)
+
+		// Create the copy of the ELF with an optional substitution
+		intOptions := make(map[string]uint32)
+		strOptions := make(map[string]string)
+		switch test.kind {
+		case symbolUint32:
+			intOptions[test.key] = test.intValue
+		case symbolString:
+			strOptions[test.key] = test.strValue
+		}
+		objectCopy := filepath.Join(tmpDir, fmt.Sprintf("%d_%s", i, elfObjCopy))
+		err = elf.Write(objectCopy, intOptions, strOptions)
+		c.Assert(err, test.elfValid)
+		if test.elfValid == notValidOptions {
+			continue
+		}
+		err = compareFiles(baseObjPath, objectCopy)
+		c.Assert(err, Equals, test.elfChangeErr)
+
+		// Test that the written ELF matches expectations
+		modifiedElf, err := Open(objectCopy)
+		c.Assert(err, IsNil)
+		defer modifiedElf.Close()
+
+		switch test.kind {
+		case symbolUint32:
+			value, err := modifiedElf.readOption(test.key)
+			c.Assert(err, IsNil)
+			c.Assert(value, Equals, test.intValue)
+		case symbolString:
+			err := modifiedElf.findString(test.strValue)
+			c.Assert(err, IsNil)
+		default:
+			_, err = modifiedElf.readOption("unknown")
+			c.Assert(err, NotNil)
+		}
+		modifiedElf.Close()
+	}
+}
+
+// BenchmarkWriteELF benchmarks writing a very simple elf demo program.
+func BenchmarkWriteELF(b *testing.B) {
+	tmpDir, err := ioutil.TempDir("", "cilium_")
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	elf, err := Open(baseObjPath)
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer elf.Close()
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		intOptions := make(map[string]uint32)
```

Comment:
Wouldn't make sense to stress test with non-empty `intOptions` and `strOptions`, as `elf.copy(...)` returns early if both are empty?

Reply:
Yep, for a more realistic benchmark the PR #7095 adds a full endpoint ELF write benchmark to pkg/data path/loader :)